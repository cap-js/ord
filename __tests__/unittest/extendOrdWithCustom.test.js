const path = require('path');
const { extendCustomORDContentIfExists } = require('../../lib/extendOrdWithCustom');

jest.mock('path', () => ({
    join: jest.fn(),
}));

jest.mock("@sap/cds", () => {
    const actualCds = jest.requireActual('@sap/cds');
    return {
        ...actualCds,
        env: {},
        log: jest.fn(() => ({
            warn: jest.fn(),
        })),
    };
});

const cds = require("@sap/cds");

describe('extendOrdWithCustom', () => {
    let appConfig = {};
    let logger;

    beforeEach(() => {
        logger = { warn: jest.fn() };
        appConfig = {
            env: {
                customOrdContentFile: 'customOrdContentFile.json',
            },
        };
    });

    afterEach(() => {
        jest.resetModules();
    });

    describe('extendCustomORDContentIfExists', () => {
        it('should skip if there is no custom ord file', () => {
            const ordContent = {};
            appConfig.env.customOrdContentFile = undefined;
            const result = extendCustomORDContentIfExists(appConfig, ordContent);
            expect(result).toEqual(ordContent);
        });

        it('should ignore and log warn if found ord top-level primitive property in customOrdFile', () => {
            const ordContent = {};
            prepareTestEnvironment({ namespace: "sap.sample" }, appConfig, 'testCustomORDContentFileThrowErrors.json');
            const result = extendCustomORDContentIfExists(appConfig, ordContent, logger);
            expect(logger.warn).toHaveBeenCalledTimes(3);
            expect(result).toMatchSnapshot();
        });

        it('should add new ord resources that are not supported by cap framework', () => {
            const ordContent = {};
            prepareTestEnvironment({}, appConfig, 'testCustomORDContentFileWithNewResources.json');
            const result = extendCustomORDContentIfExists(appConfig, ordContent);
            expect(result).toMatchSnapshot();
        });

        it('should enhance the list of generated ord resources', () => {
            const ordContent = { packages: [{ ordId: "sap.sm:package:smDataProducts:v1", localId: "smDataProductsV1" }] };
            prepareTestEnvironment({}, appConfig, 'testCustomORDContentFileWithEnhanced.json');
            const result = extendCustomORDContentIfExists(appConfig, ordContent);
            expect(result).toMatchSnapshot();
        });

        it('should should patch the existing generated ord resources', () => {
            const ordContent = {
                packages: [{
                    ordId: "sap.sm:package:smDataProducts:v1",
                    localId: "smDataProductsV1"
                }],
                apiResources: [{
                    ordId: "sap.sm:apiResource:SupplierService:v1",
                    title: "should be removed",
                    partOfGroups: [
                        "sap.cds:service:sap.test.cdsrc.sample:originalService"
                    ],
                    partOfPackage: "sap.sm:package:smDataProducts:v2",
                    extensible: {
                        "supported": "no"
                    },
                    entityTypeMappings: [
                        {
                            entityTypeTargets: [
                                {
                                    ordId: "sap.odm:entityType:BusinessPartner:v2"
                                },
                                {
                                    ordId: "sap.odm:entityType:BusinessPartner:v3"
                                }
                            ]
                        }
                    ]
                },
                {
                    ordId: "sap.sm:apiResource:orginalService:v2",
                    partOfGroups: [
                        "sap.cds:service:sap.test.cdsrc.sample:originalService"
                    ],
                    partOfPackage: "sap.sm:package:smDataProducts:v2",
                    entityTypeMappings: [
                        {
                            entityTypeTargets: []
                        }
                    ]
                }]
            };
            prepareTestEnvironment({}, appConfig, 'testCustomORDContentFileWithPatch.json');
            const result = extendCustomORDContentIfExists(appConfig, ordContent);
            expect(result).toMatchSnapshot();
        });
    });
});

function prepareTestEnvironment(ordEnvVariables, appConfig, testFileName) {
    cds.env["ord"] = ordEnvVariables;
    appConfig.env.customOrdContentFile = testFileName;
    path.join.mockReturnValue(`${__dirname}/utils/${testFileName}`);
}
