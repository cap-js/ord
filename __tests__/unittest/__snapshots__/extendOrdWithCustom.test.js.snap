// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extendOrdWithCustom extendCustomORDContentIfExists should add new ord resources that are not supported by cap framework 1`] = `
{
  "dataProducts": [
    {
      "category": "business-object",
      "description": "The Supplier data product offers access to all customers.",
      "entityTypes": [
        "sap.odm:entityType:BusinessPartner:v1",
        "sap.sm:entityType:BusinessPartner:v1",
      ],
      "lastUpdate": "2024-06-20T14:04:01+01:00",
      "localId": "Supplier",
      "ordId": "sap.sm:dataProduct:Supplier:v1",
      "outputPorts": [
        {
          "ordId": "sap.sm:apiResource:SupplierService:v1",
        },
      ],
      "partOfPackage": "sap.sm:package:smDataProducts:v1",
      "releaseStatus": "active",
      "responsible": "sap:ach:CIC-DP-CO",
      "shortDescription": "Ariba Supplier data product",
      "title": "Supplier",
      "type": "primary",
      "version": "1.1.11",
      "visibility": "public",
    },
  ],
}
`;

exports[`extendOrdWithCustom extendCustomORDContentIfExists should enhance the list of generated ord resources 1`] = `
{
  "packages": [
    {
      "localId": "smDataProductsV1",
      "ordId": "sap.sm:package:smDataProducts:v1",
    },
    {
      "localId": "smDataProductsV2",
      "ordId": "sap.sm:package:smDataProducts:v2",
    },
  ],
}
`;

exports[`extendOrdWithCustom extendCustomORDContentIfExists should ignore and log warn if found ord top-level primitive property in customOrdFile 1`] = `
{
  "packages": [
    {
      "localId": "smDataProducts",
      "ordId": "sap.sm:package:smDataProducts:v1",
    },
  ],
}
`;

exports[`extendOrdWithCustom extendCustomORDContentIfExists should should patch the existing generated ord resources 1`] = `
{
  "apiResources": [
    {
      "entityTypeMappings": [
        {
          "entityTypeTargets": [
            {
              "ordId": "sap.odm:entityType:BusinessPartner:v1",
            },
          ],
        },
      ],
      "entryPoints": [
        "/odata/v4/admin",
      ],
      "ordId": "sap.sm:apiResource:SupplierService:v1",
      "partOfGroups": [
        "sap.cds:service:sap.test.cdsrc.sample:AdminService",
      ],
      "partOfPackage": "sap.sm:package:smDataProducts:v1",
    },
    {
      "entityTypeMappings": [
        {
          "entityTypeTargets": [],
        },
      ],
      "ordId": "sap.sm:apiResource:orginalService:v2",
      "partOfGroups": [
        "sap.cds:service:sap.test.cdsrc.sample:originalService",
      ],
      "partOfPackage": "sap.sm:package:smDataProducts:v2",
    },
  ],
  "packages": [
    {
      "localId": "overrideSmDataProductsV1",
      "ordId": "sap.sm:package:smDataProducts:v1",
      "version": "1.1.1",
    },
    {
      "localId": "smDataProductsV2",
      "ordId": "sap.sm:package:smDataProducts:v2",
    },
  ],
}
`;
